message_ix_models/util/__init__.py

```diff
@@ -161,14 +161,29 @@ def aggregate_codes(df: pd.DataFrame, dim: str, codes):  # pragma: no cover
         print(key, group_series.replace({dim: mapping}))
 
 
-def broadcast(df, **kwargs):
+def broadcast(df, *args, **kwargs):
     """Fill missing data in `df` by broadcasting.
 
     Arguments
     ---------
+    args
+        tuble/dataframe mapping multiple dimensions to be fill.
     kwargs
         Keys are dimensions. Values are labels along that dimension to fill.
+ 
     """
+    labels = pd.DataFrame(*args)
+    if not labels.empty:
+        for cc in labels.columns:
+            if cc in df.columns:
+                assert df[cc].isna().all(), f"Dimension {dim} was not empty\n\n{df.head()}"
+                df.drop(cc, axis=1,inplace = True)
+                # add dummy common variable between df and labels
+        df['tmp'] = 1
+        labels['tmp'] = 1
+        df = pd.merge(df,labels).drop(columns ={'tmp'})
```

Comment:
inplace=True in pandas can be expensive/slow, so perhaps something like this:

```python
dfs = [ ]  # Will be a list of pd.DataFrame

# iterrows() produces tuples of (row index, row as pd.Series); use "_" to throw away the index, which is unused
for _, row in labels.iterrows():

    dfs.append(
        # row is like pd.Series({"year_vtg": 2010, "year_act": 2020}) so we can pass it to assign() as keyword args
        df.assign(**row)
    )

# Recombine to a single pd.DataFrame
df = pd.concat(dfs, ignore_index=True)
```

Or using a list comprehension (inline loop):
```python
df = pd.concat([df.assign(**row) for _, row in labels.iterrows()], ignore_index=True)
```